src/platform/tls_openssl.c

```diff
@@ -0,0 +1,3437 @@
+// allocated Server Name Indication (SNI) data. It is safe to call with
+// a NULL pointer.
+//
+// This function should be called once a TLS context is no longer in use
+// to avoid memory leaks.
+//
+// @param[in] TlsContext
+//     Pointer to the TLS context to clean up. May be NULL.
+//
+// @note After calling this function, the pointer must not be used again.
+//
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+CxPlatTlsUninitialize(
+    _In_opt_ CXPLAT_TLS* TlsContext
+    )
+{
+    if (TlsContext != NULL) {
+        QuicTraceLogConnVerbose(
+            OpenSslContextCleaningUp,
+            TlsContext->Connection,
+            "Cleaning up");
+
+        if (TlsContext->SNI != NULL) {
+            CXPLAT_FREE(TlsContext->SNI, QUIC_POOL_TLS_SNI);
+            TlsContext->SNI = NULL;
+        }
+
+        if (TlsContext->Ssl != NULL) {
+            SSL_free(TlsContext->Ssl);
+            TlsContext->Ssl = NULL;
+        }
+
+        CXPLAT_FREE(TlsContext, QUIC_POOL_TLS_CTX);
+    }
+}
+
+//
+// @brief Updates the HKDF label set used by the TLS context.
+//
+// This function assigns a new set of HKDF (HMAC-based Key Derivation Function)
+// labels to the given TLS context. These labels are used during QUIC's
+// cryptographic key derivation process as part of the TLS handshake.
+//
+// @param[in] TlsContext
+//     Pointer to the TLS context to update.
+//
+// @param[in] Labels
+//     Pointer to the new HKDF labels to assign. Must remain valid for the
+//     lifetime of the TLS context or until replaced.
+//
+// @note This function does not validate or duplicate the label data. The
+//       caller must ensure the pointer remains valid.
+//
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+CxPlatTlsUpdateHkdfLabels(
+    _In_ CXPLAT_TLS* TlsContext,
+    _In_ const QUIC_HKDF_LABELS* const Labels
+    )
+{
+    TlsContext->HkdfLabels = Labels;
+}
+
+//
+// @brief Allocates and initializes a new TLS record entry.
+//
+// This function creates a new `record_entry` structure, allocates memory
+// for a private copy of the given TLS record, and stores its length and
+// associated SSL context. The returned structure is used to buffer TLS
+// records for later transmission or processing in QUIC.
+//
+// @param[in] record
+//     Pointer to the TLS record data to copy.
+//
+// @param[in] RecLen
+//     Length of the TLS record in bytes.
+//
+// @param[in] ssl
+//     Pointer to the associated OpenSSL `SSL` context.
+//
+// @return
+//     A pointer to the newly allocated `record_entry` on success, or NULL
+//     if memory allocation fails.
+//
+// @note The caller is responsible for freeing the returned structure and
+//       its `record` buffer when no longer needed.
+//
+static RECORD_ENTRY *MakeNewRecord(const uint8_t *Record, size_t RecLen, SSL *Ssl)
+{
+    RECORD_ENTRY *new;
+
+    //
+    // Allocate a new structure, make sure its zeroed out
+    //
+    new = calloc(1, sizeof(RECORD_ENTRY));
```

Comment:
I think this should use `CXPLAT_ALLOC_NONPAGED`, and also create a new tag to track these allocations.